File: Algorithm/c++/1717_집합의표현.cpp
#include<iostream>

#define MAX_NODE 1000001

using namespace std;

int arr[MAX_NODE];

int findParent(const int num) {
	if (arr[num] == num) return num;
	return arr[num] = findParent(arr[num]);
}

void merge(const int a, const int b) {
	int x = findParent(a);
	int y = findParent(b);

	if (x != y) {
		arr[y] = x;
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int n, m;
	cin >> n >> m;

	for (register int i = 1; i <= n; ++i)
		arr[i] = i;

	while (m--) {
		int div, num1, num2;
		cin >> div >> num1 >> num2;

		if (div == 0) {
			merge(num1, num2);
		}
		else {
			int a = findParent(num1);
			int b = findParent(num2);

			if (a == b) {
				cout << "yes\n";
			}
			else
				cout << "no\n";
		}
	
	}

}

File: Algorithm/python/1717_집합의표현.py
import sys

sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline

n, m = map(int, input().split())
number = [i for i in range(n + 1)]


def find(num):
    if number[num] == num:
        return num
    else:
        number[num] = find(number[num])
        return number[num]


def merge(num1, num2):
    x = find(num1)
    y = find(num2)

    if x != y:
        number[y] = x


for a in range(m):
    idx, num1, num2 = map(int, input().split())
    if idx == 0:
        merge(num1, num2)
    else:
        x = find(num1)
        y = find(num2)

        if x == y:
            print("yes")
        else:
            print("no")
